### ISSUE TITLE
--rotate-pages yielding wrong results, do not use full-scale image for osd

### ISSUE THREAD

**Issue by oblakhh** [jbarlow83/OCRmyPDF#281]
For some files the `--rotate-pages` switch does not yield the desired result. Pages are left in their original wrong orientataion.

Affected Version: 7.0.0, installed via Homebrew on macOS 10.13.6, Tesseract 3.05.01.

Story:
A particular file should be rotated by 90° to the right, but OCRmyPDF does not recognize it:

```
$ ocrmypdf --rotate-pages test.pdf test-out.pdf
...
INFO -    1: page is facing ⇧, confidence 0.99 - no change
```

Using the `-k` to get access to the intermediate files, and presenting the extracted page image to tesseract directly will result in a solid confidence value:

```
$ tesseract -psm 0 test/000001.page.png -
...
Page number: 0
Orientation in degrees: 270
Rotate: 90
Orientation confidence: 13.61
Script: Latin
Script confidence: 19.11
```

Unfortunately I do not (yet) have a public test file for you for data protection reasons, I am trying to produce a solid example at the moment. In the meantime I crawled the code a little and I think I found the source of my problems: Seemingly OCRmyPDF uses a small scale **preview** image of the page to detect the orientation:

```
# _pipeline.py:365
def orient_page(
        infiles,
        output_file,
        log,
        context):

# ...
    preview = next(ii for ii in infiles if ii.endswith('.preview.jpg'))

    orient_conf = tesseract.get_orientation(
        preview,
        engine_mode=options.tesseract_oem,
        timeout=options.tesseract_timeout,
        log=log)
```

Presenting this preview file to tesseract, yields the same unwanted result:

```
$ tesseract -psm 0 test/000001.ocr.preview.jpg -
Warning. Invalid resolution 0 dpi. Using 70 instead.
Page number: 0
Orientation in degrees: 0
Rotate: 0
Orientation confidence: 0.89
Script: Latin
Script confidence: 0.97
```

The numbers don't match perfectly for unknown reason but it's close enough for me to raise my eye brow.

> I am not sure if the dpi warning is relevant here, anyway, I can't change the originating files, so I have to live with that.

I would love to change the code and try it for myself, but as I have zero python experience and am still struggling to set up a working dev environment, I will keep trying.

As a result of these (hopefully valid) findings, my short-term change request is:

* Will it be possible to introduce a switch that leads to using of the full-size image to do the orientation test, like `--rotate-pages-use-full-resolution`?

Thanks for the great work and sorry for the missing sample, I am still trying to produce one. If you have any docs to setting up a valid dev env, this would be very welcome.

**Comment by oblakhh** [comment]
Update: As my files are mostly at 200 dpi, setting`--oversample 300` does have a positive effect, but still not good enough (confidence about 5.0).

When I change `_pipeline.py` at line 356 and 357 from
```
canvas_dpi = get_canvas_square_dpi(pageinfo, options) / 2
page_dpi = get_page_square_dpi(pageinfo, options) / 2
```
to
```
canvas_dpi = get_canvas_square_dpi(pageinfo, options) 
page_dpi = get_page_square_dpi(pageinfo, options) 
```
I do get much better results (at the cost of slower processing).

Now, what I ask to myself, could we not move osd to the very end of the pipeline after all image optimizations have taken place? We would then not need the preview at all and could use the optimized image for osd.

**Comment by jbarlow83** [comment]
For dev env see https://ocrmypdf.readthedocs.io/en/latest/installation.html#installing-head-revision-from-sources (with the caveat that I changed this page recently and the link may break). Also `pip install -r *requirements.txt`.

**Comment by jbarlow83** [comment]
Deskew and clean work best when the orientation is corrected first, but orientation detection works better on a cleaned and deskewed image: it's circular. I do orientation detection first because I believe it is more robust (Tesseract can deskew and clean the input image however it likes).

It would be possible, however, to remove the preview image entirely from the pipeline and work directly with the main image. This would save a little over your version in that it avoids rasterizing the page twice, but orientation detection is going to be more expensive overall on a larger image. I'll do some testing....

See also #279.

**Comment by jbarlow83** [comment]
Confirmed that DPI does adversely affect quality too much on marginal scans/small font sizes. Addressed in v7.0.2.

For now I'm keeping the pipeline as is and letting `--rotate-pages` be a bit more expensive. I'd certainly welcome a PR if you want to tackle eliminating the duplicate raster or experiment with running orientation detection after image processing. A rough sketch would be to make `rasterize_preview` decide to how to rasterize like `rasterize_with_ghostscript`, and except with rotating, pass that to `orient_page`, and then replace the existing `rasterize_with_ghostscript` with a step that just rotates the image.

**Comment by oblakhh** [comment]
Thanks for resolving the issue at lightning speed! Updated today and results are great again.
I am definitely willing to work on the issue as you proposed. 

> I'd certainly welcome a PR if you want to tackle eliminating the duplicate raster or experiment with running orientation detection after image processing

I'm already at it, but still have to make myself comfortable with the whole environment. You can expect outcome until next week.
